### PR TITLE
[TfL] Initialise single sign on

### PR DESCRIPTION
**Pull request by MorayMySoc**
Adding single sign on for TfL staff

[skip changelog]

### CODE REVIEW

**Review by chrismytton — CHANGES_REQUESTED** [review]
You'll need to add a `sub user_from_oidc` method in the cobrand to pull the appropriate fields out of the id_token that's returned.

When I did it for Camden I also added in some (temporary) logging so I could see the fields in the id_token, see https://github.com/mysociety/fixmystreet/pull/4170. Then in https://github.com/mysociety/fixmystreet/pull/4295 I updated it to use the correct fields and removed the temporary logging.

Regarding the server config, I think the `logout_uri` needs to be an Azure URL, which we then call to log the user out in Azure. This is the code that uses the `logout_uri` config:

https://github.com/mysociety/fixmystreet/blob/0eac24d8c4e6bce7d404b13d0f6b18e9a05b6563/perllib/FixMyStreet/App/Controller/Auth/Social.pm#L219-L225

File: perllib/FixMyStreet/Cobrand/TfL.pm [hunk 1]
```diff
@@ -185,6 +185,26 @@ sub around_nearby_filter {
...
+sub user_from_oidc {
+    my ($self, $payload) = @_;
+
+    # TODO: Remove this once we know which fields the name/email are in.
+    use Data::Dumper;
+    $self->{c}->log->info("TfL OIDC payload: " . Dumper($payload));
+
+    # Extract the user's name and email address from the payload.
+    my $name = $payload->{first_name};
+    my $email = $payload->{last_name};
```
Comment (chrismytton) [inline perllib/FixMyStreet/Cobrand/TfL.pm:203]: Unlikely that the email is in the last_name field, but I guess doesn't really matter if you're coming back to change this anyway! 🙂

**Review by chrismytton — APPROVED** [review]
Looks good! 👍

**Comment by chrismytton** [comment]
@MorayMySoc Can this be closed now https://github.com/mysociety/fixmystreet/pull/4796 has been merged?

**Comment by MorayMySoc** [comment]
As above - this will not be merged as was superseded by https://github.com/mysociety/fixmystreet/pull/4796

**Pull request closed without merging**